src/Analyzers/Core/Analyzers/NamespaceSync/AbstractNamespaceSyncDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,171 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Analyzers.NamespaceSync
+{
+    internal abstract class AbstractNamespaceSyncDiagnosticAnalyzer<TNamespaceSyntax> : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+        where TNamespaceSyntax : SyntaxNode
+    {
+        public const string RootNamespaceOption = "build_property.RootNamespace";
+        public const string ProjectDirOption = "build_property.ProjectDir";
+
+        public AbstractNamespaceSyncDiagnosticAnalyzer(LocalizableResourceString title, LocalizableResourceString message)
+            : base(IDEDiagnosticIds.NamespaceSyncAnalyzerDiagnosticId,
+                EnforceOnBuild.Recommended,
+                (IPerLanguageOption?)null,
+                title,
+                message)
+        {
+        }
+
+        /// <summary>
+        /// Gets the language specific syntax facts
+        /// </summary>
+        protected abstract ISyntaxFacts GetSyntaxFacts();
+
+        /// <summary>
+        /// Returns true if the namespace declaration contains one or more partial types with multiple declarations.
+        /// </summary>
+        /// <returns></returns>
+        protected abstract bool ContainsPartialTypeWithMultipleDeclarations(TNamespaceSyntax namespaceDeclaration, SemanticModel semanticModel);
+
+        /// <summary>
+        /// Gets the syntax representing the name of a namespace declaration
+        /// </summary>
+        protected abstract SyntaxNode GetNameSyntax(TNamespaceSyntax namespaceDeclaration);
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected void AnalyzeNamespaceNode(SyntaxNodeAnalysisContext context)
+        {
+            // It's ok to not have a rootnamespace property, but if it's there we want to use it correctly
+            context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(RootNamespaceOption, out var rootNamespace);
+
+            // Project directory is a must to correctly get the relative path and construct a namespace
+            if (!context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(ProjectDirOption, out var projectDir)
+                || string.IsNullOrEmpty(projectDir))
+            {
+                return;
+            }
+
+            if (context.Node is TNamespaceSyntax namespaceDecl)
+            {
+                var currentNamespace = GetNamespaceName(namespaceDecl, rootNamespace);
```

Comment:
nit: consider remove this and add another out param `currentNamespace to `IsFileAndNamespaceMismatch` `.

Reply:
I thought about that, and it just looked weird to me. I moved GetNamespaceName to a local function and passed `currentNamespace` in instead. Hopefully that helps a bit.